### ISSUE TITLE
Implement Bazel module lockfile support

### ISSUE THREAD

**Issue by rarkins** [renovatebot/renovate#25557]
### Discussed in https://github.com/renovatebot/renovate/discussions/25532

<div type='discussions-op-text'>

<sup>Originally posted by **cgrindel** November  1, 2023</sup>
### Requirements

- Read the `.bazelversion`. Ensure that it is 6.4.0 or later.
- Read the bazelrc files looking for `--lockfile_mode`. If not found, it is assumed to be `update`.
- Lockfile Mode
  - `update`: update the lockfile. 
  - `error`: do not update the lockfile. NOTE: It is unclear why one would configure Renovate to update Bazel and have their lockfile mode be `error`.
  - `off`: do not update the lockfile.
- If a lockfile update is warranted, execute `bazel mod deps`.
- Need to ensure that [bazelisk](https://github.com/bazelbuild/bazelisk/) is installed on the platforms that execute Renovate (e.g., Docker images).

### Related Links

- [Bazel module lockfile documentation](https://bazel.build/external/lockfile)
- [Lockfile maintenance config](https://github.com/renovatebot/renovate/blob/5d9296e0b9ce5dbd503871f7d9ca5a8cca4eab83/docs/usage/configuration-options.md#lockfilemaintenance)
- [Lockfile maintenance schedule](https://github.com/renovatebot/renovate/blob/837c8850d848e32c056a0c42e9c1a756a6e7b4eb/docs/usage/faq.md#keep-lock-files-including-sub-dependencies-up-to-date-even-when-packagejson-hasnt-changed)</div>

**Comment by lalten** [comment]
Any update on this? Does somebody have a workaround maybe?

**Comment by rarkins** [comment]
A reproduction repo is highly desirable for this issue. See instructions below:

**Comment by cameron-martin** [comment]
Here's a repro: https://github.com/cameron-martin/renovate-module-bazel-lock-repro

**Comment by rarkins** [comment]
Thanks, reproduction forked to https://github.com/renovate-reproductions/25557

**Comment by rarkins** [comment]
`bazelisk` is already supported in Renovate. I think this is all the prerequisite that's needed:

```
❯ docker run --rm -it renovate/renovate:39 /bin/bash
ubuntu@6eedbf17ddd3:/usr/src/app$ install-tool bazelisk 1.25.0
[07:41:49.508] INFO (11): Installing tool bazelisk@1.25.0...
2024/12/15 07:41:53 Downloading https://releases.bazel.build/8.0.0/release/bazel-8.0.0-linux-x86_64...
Downloading: 61 MB out of 61 MB (100%)
Bazelisk version: v1.25.0
WARNING: Invoking Bazel in batch mode since it is not invoked from within a workspace (below a directory having a MODULE.bazel file).
Extracting Bazel installation...
OpenJDK 64-Bit Server VM warning: Options -Xverify:none and -noverify were deprecated in JDK 13 and will likely be removed in a future release.
Build label: 8.0.0
Build target: @@//src/main/java/com/google/devtools/build/lib/bazel:BazelServer
Build time: Mon Dec 09 18:00:43 2024 (1733767243)
Build timestamp: 1733767243
Build timestamp as int: 1733767243
[07:42:24.127] INFO (11): Install tool bazelisk succeeded in 34.6s.
ubuntu@6eedbf17ddd3:/usr/src/app$ bazel --version
bazel 8.0.0
```

**Comment by rarkins** [comment]
In this case then it looks like the remaining thing is to implement `updateArtifacts()` for the `bazel-modules` manager: https://github.com/renovatebot/renovate/blob/d0c0aa7333b0c00582dfc8cc2dbebe7e4f80d0e7/docs/development/adding-a-package-manager.md?plain=1#L89-L99

**Comment by cameron-martin** [comment]
I think the `bazelisk` manager will also need this, since updating the bazel version can change the `MODULE.bazel.lock`.

**Comment by rarkins** [comment]
How can renovate control which version of "bazel" is installed when we install bazelisk?

**Comment by cameron-martin** [comment]
I don't think you can control the version of bazel when you install bazelisk. Instead, bazelisk lazily downloads the correct version of bazel when invoked, based on [various criteria](https://github.com/bazelbuild/bazelisk?tab=readme-ov-file#how-does-bazelisk-know-which-bazel-version-to-run), .e.g the `.bazelversion` file in the repo.

**Comment by rarkins** [comment]
Important to know if that's at the time of "install-tool bazelisk" or "bazel update". For our -full images we want all tools preinstalled (on latest versions)

**Comment by cameron-martin** [comment]
Downloading a specific bazel version happens at the time of invoking `bazel`, via the bazelisk wrapper. If you want the latest version of bazel pre-cached in the docker image, you can run some bazel command such as `USE_BAZEL_VERSION=latest bazel --version` during creation of the docker image.

**Comment by joca-bt** [comment]
I think that renovate doesn't need to scan for `--lockfile_mode`. When renovate runs `bazel mod deps`, the command will respect whatever is configured in the project's `.bazelrc`.

> If a lockfile update is warranted, execute bazel mod deps.

An update will be warranted if there are updates from:
- manager bazelisk
- any datasources from manager bazel except maven, as for maven another lockfile is used

> How can renovate control which version of "bazel" is installed when we install bazelisk?

Why does renovate need to know? If bazel's version is updated then wouldn't running `bazel mod deps` after the file is updated be enough?

**Comment by cameron-martin** [comment]
When I try to use this with the public renovate app, I get an error:

```
Post-upgrade command 'bazel mod deps' has not been added to the allowed list in allowedCommands
```

**Comment by rarkins** [comment]
Running arbitrary commands is only supported if either:
1. You self-hosted, and allow such commands in config, or
2. You are a Renovate Enterprise Cloud customer

For security reasons we do not allow free users to run arbitrary commands in the free Renovate Community Cloud plan

**Comment by tacascer** [comment]
Hi!

I would very like this feature. Is there ongoing work for this? Should I submit a PR?

**Comment by tacascer** [comment]
Not sure there's more work to be done for this issue. The lock file support seems to be working (e.g. https://github.com/LowkeyLab/gradle-build-scan-server/pull/85/changes)
